pkg/connector/tickets.go

```diff
@@ -0,0 +1,449 @@
+package connector
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"path"
+	"strconv"
+	"time"
+
+	v2 "github.com/conductorone/baton-sdk/pb/c1/connector/v2"
+	"github.com/conductorone/baton-sdk/pkg/annotations"
+	"github.com/conductorone/baton-sdk/pkg/pagination"
+	sdkTicket "github.com/conductorone/baton-sdk/pkg/types/ticket"
+	"github.com/grpc-ecosystem/go-grpc-middleware/logging/zap/ctxzap"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	jira "github.com/conductorone/go-jira/v2/cloud"
+)
+
+type TicketManager interface {
+	GetTicket(ctx context.Context, ticketId string) (*v2.Ticket, annotations.Annotations, error)
+	CreateTicket(ctx context.Context, ticket *v2.Ticket, schema *v2.TicketSchema) (*v2.Ticket, annotations.Annotations, error)
+	GetTicketSchema(ctx context.Context, schemaID string) (*v2.TicketSchema, annotations.Annotations, error)
+	ListTicketSchemas(ctx context.Context, pToken *pagination.Token) ([]*v2.TicketSchema, string, annotations.Annotations, error)
+}
+
+func ticketBuilder(j *Jira) TicketManager {
+	return j
+}
+
+func (j *Jira) ListTicketSchemas(ctx context.Context, p *pagination.Token) ([]*v2.TicketSchema, string, annotations.Annotations, error) {
+	var ret []*v2.TicketSchema
+
+	offset := 0
+	// get offset from page token if its not empty
+	if p != nil && p.Token != "" {
+		var err error
+		offset, err = strconv.Atoi(p.Token)
+		if err != nil {
+			return nil, "", nil, err
+		}
+	}
+
+	projects, _, err := j.client.Project.Find(ctx, jira.WithStartAt(offset), jira.WithMaxResults(p.Size), jira.WithExpand("issueTypes"))
+	if err != nil {
+		return nil, "", nil, wrapError(err, "failed to get projects")
+	}
+
+	for _, project := range projects {
+		schema, err := j.schemaForProject(ctx, project)
+		if err != nil {
+			return nil, "", nil, err
+		}
+		ret = append(ret, schema)
+	}
+
+	nextPageToken := ""
+	if len(ret) > 1 {
+		nextPageToken = fmt.Sprintf("%d", offset+len(ret))
+	}
+
+	return ret, nextPageToken, nil, nil
+}
+
+func (j *Jira) getTicketStatuses(ctx context.Context, projectId string) ([]*v2.TicketStatus, error) {
```

Comment:
Remove this or use this in list ticket schemas, probably repurpose and isolate the pagination in here